empire/config_manager.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import os
+import shutil
+import yaml
+
+from pathlib import Path
+
+log = logging.getLogger(__name__)
+
+user_home = Path.home()
+SOURCE_CONFIG_CLIENT = Path("empire/client/config.yaml")
+SOURCE_CONFIG_SERVER = Path("empire/server/config.yaml")
+CONFIG_DIR = user_home / ".empire"
+CONFIG_CLIENT_PATH = CONFIG_DIR / "client" / "config.yaml"
+CONFIG_SERVER_PATH = CONFIG_DIR / "server" / "config.yaml"
+
+def config_init():
+    CONFIG_CLIENT_PATH.parent.mkdir(parents=True, exist_ok=True)
+    CONFIG_SERVER_PATH.parent.mkdir(parents=True, exist_ok=True)
+    
+    if not CONFIG_CLIENT_PATH.exists():
+        shutil.copy(SOURCE_CONFIG_CLIENT, CONFIG_CLIENT_PATH)
+        log.info(f"Copied {SOURCE_CONFIG_CLIENT} to {CONFIG_CLIENT_PATH}")
+    else:
+        log.info(f"{CONFIG_CLIENT_PATH} already exists.")
+
+    if not CONFIG_SERVER_PATH.exists():
+        shutil.copy(SOURCE_CONFIG_SERVER, CONFIG_SERVER_PATH)
+        log.info(f"Copied {SOURCE_CONFIG_SERVER} to {CONFIG_SERVER_PATH}")
+    else:
+        log.info(f"{CONFIG_SERVER_PATH} already exists.")
+
+
+def check_config_permission(config_dict: dict, config_type: str):
```

Comment:
Instead of passing the dictionary version, could we pass the `EmpireConfig` object? The object already has the paths parsed into resolved `Path` objects.

Reply:
Currently we cannot because the original Empire code manages in different manner the `EmpireCliConfig` and `EmpireConfig` (server one). Indeed on the first one the checks are done inside the class, on the second one are done outside. Just look:
### EmpireCliConfig 
```python
class EmpireCliConfig:
    def __init__(self):
        self.yaml: dict = {}
        if "--config" in sys.argv:    <---- INSIDE THE CLASS
            location = sys.argv[sys.argv.index("--config") + 1]
            log.info(f"Loading config from {location}")
            self.set_yaml(location)
        if len(self.yaml.items()) == 0:
            log.info("Loading default config")
            self.set_yaml(config_manager.CONFIG_CLIENT_PATH)
        config_manager.check_config_permission(self.yaml, "client")

    def set_yaml(self, location: str):    <---- INSIDE THE CLASS
        try:
            with open(location) as stream:
                self.yaml = yaml.safe_load(stream)
        except yaml.YAMLError as exc:
            log.error(exc)
        except FileNotFoundError as exc:
            log.error(exc)

empire_config = EmpireCliConfig()
```

### EmpireConfig 
```python
class EmpireConfig(EmpireBaseModel):
    supress_self_cert_warning: bool = Field(
        alias="supress-self-cert-warning", default=True
    )
    api: ApiConfig | None = ApiConfig()
    starkiller: StarkillerConfig
    submodules: SubmodulesConfig
    database: DatabaseConfig
    plugins: dict[str, dict[str, str]] = {}
    directories: DirectoriesConfig
    logging: LoggingConfig
    debug: DebugConfig

    model_config = ConfigDict(extra="allow")

    def __init__(self, config_dict: dict):
        super().__init__(**config_dict)
        # For backwards compatibility
        self.yaml = config_dict


def set_yaml(location: str):   <---- OUTSIDE THE CLASS
    location = Path(location).expanduser().resolve()
    try:
        with location.open() as stream:
            return yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        log.warning(exc)
    except FileNotFoundError as exc:
        log.warning(exc)


config_dict = {}    <---- OUTSIDE THE CLASS
if "--config" in sys.argv:
    location = sys.argv[sys.argv.index("--config") + 1]
    log.info(f"Loading config from {location}")
    config_dict = set_yaml(location)
if len(config_dict.items()) == 0:
    log.info("Loading default config")
    config_dict = set_yaml(config_manager.CONFIG_SERVER_PATH)

config_manager.check_config_permission(config_dict, "server")
empire_config = EmpireConfig(config_dict)
```
Of course the best option is to uniform these two classes.

Reply:
Ah, that makes sense. It would make sense to unify these and that was the original intent, but we've since decide to deprecate the client and focus development on Starkiller. So this seems okay

Reply:
Ok so here I won't touch anything.